File: src/main/java/com/nexuscommerce/cart/service/CartService.java
package com.nexuscommerce.cart.service;

import com.nexuscommerce.cart.dto.CartItemRequest;
import com.nexuscommerce.cart.dto.CartItemResponse;
import com.nexuscommerce.cart.dto.CartItemUpdateRequest;
import com.nexuscommerce.cart.dto.CartResponse;
import com.nexuscommerce.cart.entity.Cart;
import com.nexuscommerce.cart.entity.CartItem;
import com.nexuscommerce.cart.exception.CartItemNotFoundException;
import com.nexuscommerce.cart.exception.InsufficientStockException;
import com.nexuscommerce.cart.exception.ProductNotAvailableException;
import com.nexuscommerce.cart.repository.CartItemRepository;
import com.nexuscommerce.cart.repository.CartRepository;
import com.nexuscommerce.product.entity.Product;
import com.nexuscommerce.product.entity.ProductStatus;
import com.nexuscommerce.product.repository.ProductRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.function.Function;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Transactional
public class CartService {

    private final CartRepository cartRepository;
    private final CartItemRepository cartItemRepository;
    private final CartCreator cartCreator;
    private final ProductRepository productRepository;

    /**
     * Fetch (lazily creating) the caller's cart. Not read-only: the first GET
     * creates the cart row, and a write inside a read-only transaction is never
     * flushed — the previous version of this method "created" carts that didn't
     * survive the request.
     */
    public CartResponse getCart(UUID userId) {
        Cart cart = getOrCreateCart(userId);
        return buildCartResponse(cart);
    }

    public CartResponse addItem(UUID userId, CartItemRequest request) {
        Product product = validateProduct(request.productId());
        Cart cart = getOrCreateCart(userId);

        cartItemRepository.findByCartIdAndProductIdAndDeletedFalse(cart.getId(), request.productId())
                .ifPresentOrElse(existing -> {
                    int newQty = existing.getQuantity() + request.quantity();
                    validateStock(product, newQty);
                    existing.setQuantity(newQty);
                }, () -> {
                    validateStock(product, request.quantity());
                    addNewLine(cart, request);
                });

        return buildCartResponse(cart);
    }

    public CartResponse updateItem(UUID userId, UUID productId, CartItemUpdateRequest request) {
        Product product = validateProduct(productId);
        Cart cart = requireCart(userId, productId);

        CartItem item = cartItemRepository.findByCartIdAndProductIdAndDeletedFalse(cart.getId(), productId)
                .orElseThrow(() -> new CartItemNotFoundException(productId));

        validateStock(product, request.quantity());
        item.setQuantity(request.quantity());

        return buildCartResponse(cart);
    }

    public void removeItem(UUID userId, UUID productId) {
        Cart cart = requireCart(userId, productId);
        CartItem item = cartItemRepository.findByCartIdAndProductIdAndDeletedFalse(cart.getId(), productId)
                .orElseThrow(() -> new CartItemNotFoundException(productId));
        item.setDeleted(true);
    }

    public void clearCart(UUID userId) {
        // No cart → nothing to clear. Deliberately does not create one as a side
        // effect (the previous version inserted an empty cart row on DELETE).
        cartRepository.findByUserIdAndDeletedFalse(userId).ifPresent(cart ->
                cartItemRepository.findByCartIdAndDeletedFalse(cart.getId())
                        .forEach(item -> item.setDeleted(true)));
    }

    /**
     * Adding a previously-removed product un-deletes its most recent soft-deleted
     * row (with the fresh quantity) instead of inserting another one — repeated
     * remove/re-add no longer accumulates rows, and the V8 live-row unique index
     * stays satisfiable. A brand-new product inserts; if two requests race the
     * insert, the index rejects the loser with a 409 and the client's retry
     * merges normally.
     */
    private void addNewLine(Cart cart, CartItemRequest request) {
        cartItemRepository
                .findFirstByCartIdAndProductIdAndDeletedTrueOrderByUpdatedAtDesc(cart.getId(), request.productId())
                .ifPresentOrElse(removed -> {
                    removed.setDeleted(false);
                    removed.setQuantity(request.quantity());
                }, () -> cartItemRepository.save(CartItem.builder()
                        .cart(cart)
                        .productId(request.productId())
                        .quantity(request.quantity())
                        .build()));
    }

    /**
     * The V8 unique index makes concurrent creation safe: the loser of the race
     * gets {@code null} back from {@link CartCreator} (its own small transaction
     * rolled back) and re-fetches the winner's row.
     */
    private Cart getOrCreateCart(UUID userId) {
        return cartRepository.findByUserIdAndDeletedFalse(userId)
                .orElseGet(() -> {
                    Cart created = cartCreator.create(userId);
                    return created != null
                            ? created
                            : cartRepository.findByUserIdAndDeletedFalse(userId)
                                    .orElseThrow(() -> new IllegalStateException(
                                            "Cart creation race left no live cart for user " + userId));
                });
    }

    /**
     * Mutations of existing lines never create a cart as a side effect: a user
     * with no cart cannot have the line they are trying to change, so this is
     * the same 404 as a missing line.
     */
    private Cart requireCart(UUID userId, UUID productId) {
        return cartRepository.findByUserIdAndDeletedFalse(userId)
                .orElseThrow(() -> new CartItemNotFoundException(productId));
    }

    private Product validateProduct(UUID productId) {
        Product product = productRepository.findByIdAndDeletedFalse(productId)
                .orElseThrow(() -> new ProductNotAvailableException(productId));
        if (product.getStatus() != ProductStatus.ACTIVE) {
            throw new ProductNotAvailableException(productId);
        }
        return product;
    }

    private void validateStock(Product product, int requestedQty) {
        if (requestedQty > product.getStockQuantity()) {
            throw new InsufficientStockException(product.getStockQuantity(), requestedQty);
        }
    }

    private CartResponse buildCartResponse(Cart cart) {
        List<CartItem> items = cartItemRepository.findByCartIdAndDeletedFalse(cart.getId());

        // One batch product lookup for the whole cart (was one query per line).
        Map<UUID, Product> products = productRepository
                .findByIdInAndDeletedFalse(items.stream().map(CartItem::getProductId).toList())
                .stream()
                .collect(Collectors.toMap(Product::getId, Function.identity()));

        // Soft-deleted products are excluded (no data to display).
        // Products that exist but are not ACTIVE appear with available=false.
        List<CartItemResponse> itemResponses = items.stream()
                .map(item -> {
                    Product product = products.get(item.getProductId());
                    return product != null ? buildCartItemResponse(item, product) : null;
                })
                .filter(r -> r != null)
                .toList();

        // Only available items count toward totals
        BigDecimal totalPrice = itemResponses.stream()
                .filter(CartItemResponse::available)
                .map(CartItemResponse::subtotal)
                .reduce(BigDecimal.ZERO, BigDecimal::add);

        int totalItems = itemResponses.stream()
                .filter(CartItemResponse::available)
                .mapToInt(CartItemResponse::quantity)
                .sum();

        return new CartResponse(
                cart.getId(),
                cart.getUserId(),
                itemResponses,
                totalItems,
                totalPrice,
                cart.getUpdatedAt()
        );
    }

    private CartItemResponse buildCartItemResponse(CartItem item, Product product) {
        boolean available = product.getStatus() == ProductStatus.ACTIVE;
        String primaryImageUrl = product.getPrimaryImageUrl();
        BigDecimal subtotal = available
                ? product.getPrice().multiply(BigDecimal.valueOf(item.getQuantity()))
                : BigDecimal.ZERO;
        return new CartItemResponse(
                product.getId(),
                product.getName(),
                product.getSku(),
                primaryImageUrl,
                product.getPrice(),
                item.getQuantity(),
                subtotal,
                available
        );
    }
}

File: src/cart/cart.service.ts
import { Prisma, type Cart, type PrismaClient } from '@prisma/client';
import { z } from 'zod';
import { money } from '../common/api-response';
import { CartItemNotFoundError, InsufficientStockError, ProductNotAvailableError } from '../common/errors';
import { isIntegrityViolation } from '../common/error-handler';
import { integer, requiredUuid } from '../common/validation';
import { primaryImageUrl } from '../product/product.mapper';

export const CartItemSchema = z.object({
  productId: requiredUuid(),
  quantity: integer({ required: true, positive: true, max: 999 }),
});

export const CartItemUpdateSchema = z.object({
  quantity: integer({ required: true, positive: true, max: 999 }),
});

/**
 * The caller's cart. One live cart per user and one live line per product are
 * enforced by partial unique indexes (V8); the code below leans on them to make
 * concurrent requests safe rather than locking.
 */
export class CartService {
  constructor(private readonly prisma: PrismaClient) {}

  /** Fetch (lazily creating) the caller's cart. */
  async getCart(userId: string) {
    return this.buildResponse(await this.getOrCreateCart(userId));
  }

  async addItem(userId: string, input: z.output<typeof CartItemSchema>) {
    const product = await this.validateProduct(input.productId);
    const cart = await this.getOrCreateCart(userId);

    const existing = await this.prisma.cartItem.findFirst({
      where: { cartId: cart.id, productId: input.productId, deleted: false },
    });
    if (existing) {
      const newQty = existing.quantity + input.quantity;
      validateStock(product.stockQuantity, newQty);
      await this.prisma.cartItem.update({ where: { id: existing.id }, data: { quantity: newQty } });
    } else {
      validateStock(product.stockQuantity, input.quantity);
      await this.addNewLine(cart.id, input.productId, input.quantity);
    }
    return this.buildResponse(cart);
  }

  async updateItem(userId: string, productId: string, quantity: number) {
    const product = await this.validateProduct(productId);
    const cart = await this.requireCart(userId, productId);
    const item = await this.prisma.cartItem.findFirst({ where: { cartId: cart.id, productId, deleted: false } });
    if (!item) throw new CartItemNotFoundError(productId);

    validateStock(product.stockQuantity, quantity);
    await this.prisma.cartItem.update({ where: { id: item.id }, data: { quantity } });
    return this.buildResponse(cart);
  }

  async removeItem(userId: string, productId: string) {
    const cart = await this.requireCart(userId, productId);
    const item = await this.prisma.cartItem.findFirst({ where: { cartId: cart.id, productId, deleted: false } });
    if (!item) throw new CartItemNotFoundError(productId);
    await this.prisma.cartItem.update({ where: { id: item.id }, data: { deleted: true } });
  }

  /** No cart means nothing to clear — deliberately doesn't create one. */
  async clearCart(userId: string) {
    const cart = await this.prisma.cart.findFirst({ where: { userId, deleted: false } });
    if (cart) {
      await this.prisma.cartItem.updateMany({ where: { cartId: cart.id, deleted: false }, data: { deleted: true } });
    }
  }

  /**
   * Re-adding a removed product un-deletes its most recent removed row instead
   * of inserting another, so remove/re-add cycles don't pile up rows. A
   * brand-new product inserts; if two requests race that insert, the unique
   * index rejects the loser with a 409 and the client's retry merges normally.
   */
  private async addNewLine(cartId: string, productId: string, quantity: number) {
    const removed = await this.prisma.cartItem.findFirst({
      where: { cartId, productId, deleted: true },
      orderBy: { updatedAt: 'desc' },
    });
    if (removed) {
      await this.prisma.cartItem.update({ where: { id: removed.id }, data: { deleted: false, quantity } });
    } else {
      await this.prisma.cartItem.create({ data: { cartId, productId, quantity } });
    }
  }

  /**
   * Concurrent creation is safe through uniq_carts_user_live: the loser's
   * insert fails and it re-reads the winner's cart.
   */
  private async getOrCreateCart(userId: string): Promise<Cart> {
    const existing = await this.prisma.cart.findFirst({ where: { userId, deleted: false } });
    if (existing) return existing;
    try {
      return await this.prisma.cart.create({ data: { userId } });
    } catch (e) {
      if (!isIntegrityViolation(e)) throw e;
      const winner = await this.prisma.cart.findFirst({ where: { userId, deleted: false } });
      if (!winner) throw new Error(`Cart creation race left no live cart for user ${userId}`);
      return winner;
    }
  }

  /**
   * Changing a line never creates a cart: a user without one can't have the
   * line either, so it's the same 404 as a missing line.
   */
  private async requireCart(userId: string, productId: string) {
    const cart = await this.prisma.cart.findFirst({ where: { userId, deleted: false } });
    if (!cart) throw new CartItemNotFoundError(productId);
    return cart;
  }

  private async validateProduct(productId: string) {
    const product = await this.prisma.product.findFirst({ where: { id: productId, deleted: false } });
    if (!product || product.status !== 'ACTIVE') throw new ProductNotAvailableError(productId);
    return product;
  }

  private async buildResponse(cart: Cart) {
    const items = await this.prisma.cartItem.findMany({
      where: { cartId: cart.id, deleted: false },
      orderBy: { createdAt: 'asc' },
    });
    // One batch product lookup for the whole cart.
    const products = new Map(
      (
        await this.prisma.product.findMany({
          where: { id: { in: items.map((i) => i.productId) }, deleted: false },
          include: { images: { where: { deleted: false }, orderBy: { position: 'asc' } } },
        })
      ).map((p) => [p.id, p]),
    );

    // Soft-deleted products drop out (nothing to show); inactive ones stay with
    // available=false and don't count toward the totals.
    const lines = items.flatMap((item) => {
      const product = products.get(item.productId);
      if (!product) return [];
      const available = product.status === 'ACTIVE';
      const subtotal = available ? product.price.mul(item.quantity) : new Prisma.Decimal(0);
      return [{ item, product, available, subtotal }];
    });

    const totalPrice = lines.filter((l) => l.available).reduce((sum, l) => sum.add(l.subtotal), new Prisma.Decimal(0));
    const totalItems = lines.filter((l) => l.available).reduce((sum, l) => sum + l.item.quantity, 0);

    return {
      cartId: cart.id,
      customerId: cart.userId,
      items: lines.map(({ item, product, available, subtotal }) => ({
        productId: product.id,
        productName: product.name,
        sku: product.sku,
        primaryImageUrl: primaryImageUrl(product.images),
        unitPrice: money(product.price),
        quantity: item.quantity,
        subtotal: money(subtotal),
        available,
      })),
      totalItems,
      totalPrice: money(totalPrice),
      updatedAt: cart.updatedAt,
    };
  }
}

function validateStock(available: number, requested: number) {
  if (requested > available) throw new InsufficientStockError(available, requested);
}
